app/updatemanager.go

```diff
@@ -125,9 +138,137 @@ func (u *UpdateManager) run(ctx context.Context) error {
 	return nil
 }
 
+type UpdateControlMapState struct {
+	Action           string `json:"action"`
+	OnMapExpire      string `json:"on_map_expire"`
+	OnActionExecuted string `json:"on_action_executed"`
+}
+
 type UpdateControlMap struct {
-	ID       string `json:"id"`
-	Priority int    `json:"priority"`
+	ID       string                           `json:"id"`
+	Priority int                              `json:"priority"`
+	States   map[string]UpdateControlMapState `json:"states"`
+}
+
+var UpdateControlMapStateKeyValid []string = []string{
+	"ArtifactInstall_Enter",
+	"ArtifactReboot_Enter",
+	"ArtifactCommit_Enter",
+}
+var UpdateControlMapStateActionValid []string = []string{
+	"continue",
+	"force_continue",
+	"pause",
+	"fail",
+}
+var UpdateControlMapStateOnMapExpireValid []string = []string{
+	"continue",
+	"force_continue",
+	"fail",
+}
+var UpdateControlMapStateOnActionExecutedValid []string = UpdateControlMapStateActionValid
+
+var UpdateControlMapStateActionDefault string = "continue"
+
+func UpdateControlMapStateOnMapExpireDefault(action string) string {
+	if action == "pause" {
+		return "fail"
+	}
+	return action
+}
+
+var UpdateControlMapStateOnActionExecutedDefault = UpdateControlMapStateActionDefault
+
+func (s UpdateControlMapState) Validate() error {
+	if s.Action != "" {
+		found, err := utils.ElemInSlice(UpdateControlMapStateActionValid, s.Action)
+		if err != nil {
+			return err
+		}
+		if !found {
+			return fmt.Errorf("invalid value %q", s.Action)
+		}
+	}
+
+	if s.OnMapExpire != "" {
+		found, err := utils.ElemInSlice(UpdateControlMapStateOnMapExpireValid, s.OnMapExpire)
+		if err != nil {
+			return err
+		}
+		if !found {
+			return fmt.Errorf("invalid value %q", s.OnMapExpire)
+		}
+	}
+
+	if s.OnActionExecuted != "" {
+		found, err := utils.ElemInSlice(UpdateControlMapStateOnActionExecutedValid, s.OnActionExecuted)
+		if err != nil {
+			return err
+		}
+		if !found {
+			return fmt.Errorf("invalid value %q", s.OnActionExecuted)
+		}
+	}
+
+	return nil
+}
+
+func (s *UpdateControlMapState) Sanitize() {
+	if s.OnMapExpire == UpdateControlMapStateOnMapExpireDefault(s.Action) {
```

Comment:
I'm a little unsure about these, would we not want to keep the value in, even if it is the defaults? :thinking: 
Like how do I know afterwards what the default is?

Reply:
This was a bit unclear to me from the acceptance criteria and I think it is an implementation detail that can work both ways.

Either:
1) (implemented in this PR) We remove all fields that are default, in the business logic we need to consider empty values as default.
2) We set all non-provided fields to default values (except for ID everything is optional, so it is likely that a user passes incomplete maps). Then the business logic later on does not need to check for empty values.

My reason for going for 1) is that we are going to eliminate whole maps when they are all defaults (that is explicit in the acceptance criteria) so it seemed more coherent to also eliminate (or blank) fields that were default.

I don't have an strong opinion, and if you think 2) above will result in more readable code I can change it.

Reply:
I'm leaning toward 2) I think, but feel free to do what you feel will work best here

Reply:
Reworked for 2)

@oleorhagen please have another look.